File: src/app/src/main/java/com/huawei/cardreader/java/scannedcardlist/adapter/BankCardListAdapter.java
/*
 * Copyright (c) Huawei Technologies Co., Ltd. 2020-2021. All rights reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.huawei.cardreader.java.scannedcardlist.adapter;

import android.content.Context;
import android.content.Intent;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Toast;

import androidx.annotation.NonNull;
import androidx.appcompat.app.AlertDialog;
import androidx.databinding.DataBindingUtil;
import androidx.recyclerview.widget.RecyclerView;

import com.huawei.cardreader.R;
import com.huawei.cardreader.databinding.ContactListItemBinding;
import com.huawei.cardreader.java.scannedcardlist.activity.ScannedCardListActivity;
import com.huawei.cardreader.java.scannedcardlist.model.QRcodmodel;
import com.huawei.cardreader.java.userauthentication.activity.AuthenticationPage;
import com.huawei.cardreader.java.utils.Constants;
import com.huawei.cardreader.java.utils.DataConverter;

import java.util.List;

import static com.huawei.cardreader.java.utils.Constants.ID;
import static com.huawei.cardreader.java.utils.Constants.IS_DELETED;
import static com.huawei.cardreader.java.utils.Constants.IS_DELETED_FROM_PROFILEDETAILS;
import static com.huawei.cardreader.java.utils.Constants.QRCODEIMAGE;

/**
 * The type Bank card list adapter.
 */
public class BankCardListAdapter extends RecyclerView.Adapter<BankCardListAdapter.ViewHolder> {
    private List<QRcodmodel> mListData;
    private Context mContext;
    private AlertDialog.Builder mBuilder;
    private BankAdapterCallback mBankAdapterCallback;

    /**
     * Instantiates a new Bank card list adapter.
     *
     * @param mainActivity   the main activity
     * @param qRcodmodelList the qrcodmodel list
     */
    public BankCardListAdapter(ScannedCardListActivity mainActivity, List<QRcodmodel> qRcodmodelList
    ) {
        this.mListData = qRcodmodelList;
        this.mContext = mainActivity;
        mBuilder = new AlertDialog.Builder(mContext);
    }


    @NonNull
    @Override
    public ViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        ContactListItemBinding binding = DataBindingUtil.inflate
                (LayoutInflater.from(parent.getContext()), R.layout.contact_list_item,
                        parent, false);
        return new ViewHolder(binding);
    }

    @Override
    public void onBindViewHolder(@NonNull ViewHolder holder, final int position) {
        final QRcodmodel qRcodmodel = mListData.get(position);

        if (qRcodmodel.getBusinesscardUserDetailsList().get(position).getCardType
                ().equalsIgnoreCase(Constants.BANKCARD)) {
            String maskPhno;
            if (qRcodmodel.getBusinesscardUserDetailsList().get(position).getAccountnumber() != null
                    && qRcodmodel.getBusinesscardUserDetailsList().get
                    (position).getAccountnumber().length() > 0) {
                maskPhno = maskCardNumber(DataConverter.deCodeString
                        (qRcodmodel.getBusinesscardUserDetailsList().get
                        (position).getAccountnumber()), Constants.PHONENO_MASK);
            } else {
                maskPhno = Constants.DEFAULT_MASK;
            }
            String name = maskCardNumber(DataConverter.deCodeString
                    (qRcodmodel.getBusinesscardUserDetailsList().get
                    (position).getName()), Constants.NAME_MASK);
            holder.itemRowBinding.nameTextView.setText(name);
            holder.itemRowBinding.cardtypeTextView.setText(qRcodmodel.getBusinesscardUserDetailsList
                    ().get(position).getCardType());
            holder.itemRowBinding.categoryTextView.setText(Constants.BANKCARD);
            holder.itemRowBinding.imgCard.setImageBitmap(DataConverter.stringToBitMap
                    (qRcodmodel.getBusinesscardUserDetailsList().get(position).getImage()));
            holder.itemRowBinding.phnoTextView.setText(maskPhno);
            holder.itemRowBinding.jobLabel.setVisibility(View.VISIBLE);
            holder.itemRowBinding.jobLabel.setText(R.string.expireDate);
            holder.itemRowBinding.jobTextView.setVisibility(View.VISIBLE);
            String expiry = maskCardNumber(DataConverter.deCodeString
                    (qRcodmodel.getBusinesscardUserDetailsList().get
                    (position).getExpirydate()), Constants.EXPIRY_MASK);
            holder.itemRowBinding.jobTextView.setText(expiry);
            holder.itemRowBinding.companyLabel.setVisibility(View.VISIBLE);
            holder.itemRowBinding.companyLabel.setText(R.string.type);
            holder.itemRowBinding.companyTextView.setVisibility(View.VISIBLE);
            holder.itemRowBinding.companyTextView.setText
                    (DataConverter.deCodeString(qRcodmodel.getBusinesscardUserDetailsList().get
                            (position).getBankorganization()));
            holder.itemRowBinding.phnoLabel.setText(R.string.accNumber);
        }
        holder.itemRowBinding.liItems.setOnClickListener(view -> {
            IS_DELETED_FROM_PROFILEDETAILS = false;
            Intent intent = new Intent(mContext, AuthenticationPage.class);
            intent.putExtra(ID, qRcodmodel.getBusinesscardUserDetailsList().get
                    (position).getId());
            intent.putExtra(QRCODEIMAGE, qRcodmodel.getBitmap());
            intent.putExtra(Constants.QRCODECARDTYPE,qRcodmodel.getBusinesscardUserDetailsList
                    ().get(position).getCardType());
            mContext.startActivity(intent);
        });
        holder.itemRowBinding.imgDelete.setOnClickListener(view -> showalertDialog
                (qRcodmodel, holder, position));
    }


    /**
     * Showalert dialog.
     *
     * @param qRcodmodel the q rcodmodel
     * @param holder     the holder
     * @param position   the position
     */
    public void showalertDialog(QRcodmodel qRcodmodel, ViewHolder holder, int position) {
        mBuilder.setMessage(R.string.delete_dialog)
                .setCancelable(false)
                .setPositiveButton(R.string.yes, (dialog, id) -> {
                    mBankAdapterCallback.onMethodCallback
                            (qRcodmodel.getBusinesscardUserDetailsList().get(position).getId());
                    int actualPosition = holder.getAdapterPosition();
                    mListData.remove(actualPosition);
                    notifyItemRemoved(actualPosition);
                    Toast.makeText(mContext, R.string.Deleted_the_card, Toast.LENGTH_SHORT).show();
                    IS_DELETED = mListData.isEmpty();
                })
                .setNegativeButton(R.string.no, (dialog, id) -> dialog.cancel());
        AlertDialog alert = mBuilder.create();
        alert.setTitle("");
        alert.show();
    }

    /**
     * Mask card number string.
     *
     * @param cardNumber the card number
     * @param mask       the mask
     * @return the string
     */
    public static String maskCardNumber(String cardNumber, String mask) {
        int index = 0;
        StringBuilder maskedNumber = new StringBuilder();
        for (int i = 0; i < mask.length(); i++) {
            char c = mask.charAt(i);
            if (c == '#') {
                maskedNumber.append(cardNumber.charAt(index));
                index++;
            } else if (c == 'x') {
                maskedNumber.append(c);
                index++;
            } else {
                maskedNumber.append(c);
            }
        }
        // return the masked number
        return maskedNumber.toString();
    }

    @Override
    public int getItemCount() {
        return mListData.size();
    }

    /**
     * The type View holder.
     */
    public static class ViewHolder extends RecyclerView.ViewHolder {
        private ContactListItemBinding itemRowBinding;

        /**
         * Instantiates a new View holder.
         *
         * @param itemBinding the item view
         */
        public ViewHolder(ContactListItemBinding itemBinding) {
            super(itemBinding.getRoot());
            this.itemRowBinding = itemBinding;
        }
    }

    /**
     * The interface Bank adapter callback.
     */
    public interface BankAdapterCallback {
        /**
         * On bank method callback.
         *
         * @param id the id
         */
        void onMethodCallback(int id);
    }

    /**
     * Sets bank adapter callback.
     *
     * @param mBankAdapterCallback the m bank adapter callback
     */
    public void setmBankAdapterCallback(BankAdapterCallback mBankAdapterCallback) {
        this.mBankAdapterCallback = mBankAdapterCallback;
    }
}

File: src/app/src/main/java/com/huawei/cardreader/kotlin/scannedcardlist/adapter/BankCardListAdapter.kt
/*
 * Copyright (c) Huawei Technologies Co., Ltd. 2020-2021. All rights reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.huawei.cardreader.kotlin.scannedcardlist.adapter

import android.content.Context
import android.content.DialogInterface
import android.content.Intent
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.Toast
import androidx.appcompat.app.AlertDialog
import androidx.databinding.DataBindingUtil
import androidx.recyclerview.widget.RecyclerView
import com.huawei.cardreader.R
import com.huawei.cardreader.databinding.ContactListItemBinding
import com.huawei.cardreader.kotlin.scannedcardlist.activity.ScannedCardListActivity
import com.huawei.cardreader.kotlin.scannedcardlist.model.QRcodmodel
import com.huawei.cardreader.kotlin.userauthentication.activity.AuthenticationPage
import com.huawei.cardreader.kotlin.utils.Constants
import com.huawei.cardreader.kotlin.utils.Constants.IS_DELETED
import com.huawei.cardreader.kotlin.utils.Constants.IS_DELETED_FROM_PROFILEDETAILS
import com.huawei.cardreader.kotlin.utils.DataConverter.deCodeString
import com.huawei.cardreader.kotlin.utils.DataConverter.stringToBitMap

/**
 * The type Bank card list adapter.
 */
class BankCardListAdapter(
    mainActivity: ScannedCardListActivity, private val mListData: MutableList<QRcodmodel>
) : RecyclerView.Adapter<BankCardListAdapter.ViewHolder>() {
    private val mContext: Context
    private val mBuilder: AlertDialog.Builder
    private var mBankAdapterCallback: BankAdapterCallback? =
        null

    override fun onCreateViewHolder(
        parent: ViewGroup,
        viewType: Int
    ): BankCardListAdapter.ViewHolder {
        val binding: ContactListItemBinding =
            DataBindingUtil.inflate(
                LayoutInflater.from(parent.context), R.layout.contact_list_item,
                parent, false
            )
        return ViewHolder(binding)
    }

   override fun onBindViewHolder(
        holder: ViewHolder,
        position: Int
    ) {
        val qRcodmodel = mListData[position]
        if (qRcodmodel.businesscardUserDetailsList!![position].cardType.equals(
                Constants.BANKCARD,
                ignoreCase = true
            )
        ) {
            val maskPhno: String
            maskPhno = if (qRcodmodel.businesscardUserDetailsList!![position].accountnumber != null
                && qRcodmodel.businesscardUserDetailsList!![position].accountnumber!!.length > 0
            ) {
                maskCardNumber(
                    deCodeString(
                        qRcodmodel.businesscardUserDetailsList!![position].accountnumber!!
                    ),
                    Constants.PHONENO_MASK
                )
            } else {
                Constants.DEFAULT_MASK
            }
            val name =
                maskCardNumber(
                    deCodeString(qRcodmodel.businesscardUserDetailsList!![position].name!!),
                    Constants.NAME_MASK
                )
            holder.itemRowBinding.nameTextView.text = name
            holder.itemRowBinding.cardtypeTextView.text =
                qRcodmodel.businesscardUserDetailsList!![position].cardType
            holder.itemRowBinding.categoryTextView.text = Constants.BANKCARD
            holder.itemRowBinding.imgCard.setImageBitmap(
                stringToBitMap(
                    qRcodmodel.businesscardUserDetailsList!![position].image
                )
            )
            holder.itemRowBinding.phnoTextView.text = maskPhno
            holder.itemRowBinding.jobLabel.visibility = View.VISIBLE
            holder.itemRowBinding.jobLabel.setText(R.string.expireDate)
            holder.itemRowBinding.jobTextView.visibility = View.VISIBLE
            val expiry =
                maskCardNumber(
                    deCodeString(qRcodmodel.businesscardUserDetailsList!![position].expirydate!!),
                    Constants.EXPIRY_MASK
                )
            holder.itemRowBinding.jobTextView.text = expiry
            holder.itemRowBinding.companyLabel.visibility = View.VISIBLE
            holder.itemRowBinding.companyLabel.setText(R.string.type)
            holder.itemRowBinding.companyTextView.visibility = View.VISIBLE
            holder.itemRowBinding.companyTextView.text = deCodeString(
                qRcodmodel.businesscardUserDetailsList!![position].bankorganization!!
            )
            holder.itemRowBinding.phnoLabel.setText(R.string.accNumber)
        }
        holder.itemRowBinding.liItems.setOnClickListener { view: View? ->
            IS_DELETED_FROM_PROFILEDETAILS = false
            val intent = Intent(mContext, AuthenticationPage::class.java)
            intent.putExtra(Constants.ID, qRcodmodel.businesscardUserDetailsList!![position].id)
            intent.putExtra(Constants.QRCODEIMAGE, qRcodmodel.bitmap)
            intent.putExtra(Constants.QRCODECARDTYPE,Constants.BANKCARD)
            mContext.startActivity(intent)
        }
        holder.itemRowBinding.imgDelete.setOnClickListener { view: View? ->
            showalertDialog(
                qRcodmodel,
                holder,
                position
            )
        }
    }

    /**
     * Showalert dialog.
     *
     * @param qRcodmodel the q rcodmodel
     * @param holder     the holder
     * @param position   the position
     */
    fun showalertDialog(
        qRcodmodel: QRcodmodel,
        holder: ViewHolder,
        position: Int
    ) {
        mBuilder.setMessage(R.string.delete_dialog)
            .setCancelable(false)
            .setPositiveButton(
                R.string.yes
            ) { dialog: DialogInterface?, id: Int ->
                mBankAdapterCallback!!.onMethodCallback(
                    qRcodmodel.businesscardUserDetailsList!![position].id
                )
                val actualPosition = holder.adapterPosition
                mListData.removeAt(actualPosition)
                notifyItemRemoved(actualPosition)
                Toast.makeText(mContext, R.string.Deleted_the_card, Toast.LENGTH_SHORT).show()
                IS_DELETED = mListData.isEmpty()
            }
            .setNegativeButton(
                R.string.no
            ) { dialog: DialogInterface, id: Int -> dialog.cancel() }
        val alert = mBuilder.create()
        alert.setTitle("")
        alert.show()
    }

    override fun getItemCount(): Int {
        return mListData.size
    }

    /**
     * The type View holder.
     */
    class ViewHolder
    /**
     * Instantiates a new View holder.
     *
     * @param itemBinding the item view
     */(val itemRowBinding: ContactListItemBinding) :
        RecyclerView.ViewHolder(itemRowBinding.root)

    /**
     * The interface Bank adapter callback.
     */
    interface BankAdapterCallback {
        /**
         * On bank method callback.
         *
         * @param id the id
         */
        fun onMethodCallback(id: Int)
    }

    /**
     * Sets bank adapter callback.
     *
     * @param mBankAdapterCallback the m bank adapter callback
     */
    fun setmBankAdapterCallback(mBankAdapterCallback: BankAdapterCallback?) {
        this.mBankAdapterCallback = mBankAdapterCallback
    }

    companion object {
        /**
         * Mask card number string.
         *
         * @param cardNumber the card number
         * @param mask       the mask
         * @return the string
         */
        fun maskCardNumber(cardNumber: String, mask: String): String {
            var index = 0
            val maskedNumber = StringBuilder()
            for (i in 0 until mask.length) {
                val c = mask[i]
                if (c == '#') {
                    maskedNumber.append(cardNumber[index])
                    index++
                } else if (c == 'x') {
                    maskedNumber.append(c)
                    index++
                } else {
                    maskedNumber.append(c)
                }
            }
            // return the masked number
            return maskedNumber.toString()
        }
    }

    /**
     * Instantiates a new Bank card list adapter.
     *
     * @param mainActivity   the main activity
     * @param qRcodmodelList the qrcodmodel list
     */
    init {
        mContext = mainActivity
        mBuilder = AlertDialog.Builder(mContext)
    }
}